test/e2e/metriconly/metrics_test.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_metric_only
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/node-problem-detector/pkg/util/metrics"
+	"k8s.io/node-problem-detector/test/e2e/lib/gce"
+	"k8s.io/node-problem-detector/test/e2e/lib/npd"
+
+	"github.com/avast/retry-go"
+	"github.com/onsi/ginkgo"
+	"github.com/pborman/uuid"
+)
+
+var _ = ginkgo.Describe("NPD should export Prometheus metrics for detected problems", func() {
+	var instance gce.Instance
+
+	ginkgo.BeforeEach(func() {
+		var err error
+		// TODO(xueweiz): Creating instance for each test case is slow. We should either reuse the instance
+		// between tests, or have a way to run these tests in parallel.
+		instance, err = gce.CreateInstance(
+			gce.Instance{
+				Name:           "npd-metrics-" + *image + "-" + uuid.NewUUID().String()[:8],
+				Zone:           *zone,
+				Project:        *project,
+				SshKey:         *sshKey,
+				SshUser:        *sshUser,
+				ComputeService: computeService,
+			},
+			*image,
+			*imageProject)
+		if err != nil {
+			ginkgo.Fail(fmt.Sprintf("Unable to create test instance: %v", err))
+		}
+
+		err = npd.SetupNPD(instance, *npdBuildTar)
+		if err != nil {
+			ginkgo.Fail(fmt.Sprintf("Unable to setup NPD: %v", err))
+		}
+	})
+
+	ginkgo.Context("when started on a clean instance,", func() {
+
+		ginkgo.It("NPD should export Prometheus metrics for uptime", func() {
+			var gotMetrics []metrics.Float64MetricRepresentation
+			var err error
+
+			verifyUptimeMetric := func() error {
+				gotMetrics, err = npd.FetchNPDMetrics(instance)
+				if err != nil {
+					return fmt.Errorf("Error fetching NPD metrics: %v", err)
+				}
+
+				_, err = metrics.GetFloat64Metric(gotMetrics, "host_uptime", map[string]string{}, false)
+				if err != nil {
+					return fmt.Errorf("Failed to find uptime metric: %v.\nHere is all NPD exported metrics: %v",
+						err, gotMetrics)
+				}
+				return nil
+			}
+
+			// Wait for NPD to be ready for a maximum of 120 seconds.
+			err = retry.Do(verifyUptimeMetric,
```

Comment:
I think we need some generic way of waiting for NPD coming up, put that into some function and call it after `npd.SetupNPD`.

If you think waiting for uptime metric is the easiest way to check NPD is up, then put this into that function, too.

Reply:
SG. Added `WaitForNPD()` function.

I don't want to hard code the "ready signal" (e.g. host_uptime metric) in the function. Mainly because NPD is configurable. We might test it under different configurations, which will yield different ready signals.

I'm also think about maybe we should use the systemd service property as another ready signal. But in my opinion, those are minor details, and we can improve them along the way.